src/main/java/hudson/tasks/junit/pipeline/JUnitResultsStepExecution.java

```diff
@@ -59,22 +61,45 @@ protected TestResultSummary run() throws Exception {
     /**
      * Get the stage and parallel branch start node IDs (not the body nodes) for this node, innermost first.
      * @param node A flownode.
-     * @return A nonnull, possibly empty list of stage/parallel branch start node IDs, innermost first.
+     * @return A nonnull, possibly empty list of stage/parallel branch start nodes, innermost first.
      */
     @Nonnull
-    public static List<String> getEnclosingStagesAndParallels(FlowNode node) {
-        List<String> enclosingBlocks = new ArrayList<>();
+    public static List<FlowNode> getEnclosingStagesAndParallels(FlowNode node) {
+        List<FlowNode> enclosingBlocks = new ArrayList<>();
         for (FlowNode enclosing : node.getEnclosingBlocks()) {
             if (enclosing != null && enclosing.getAction(LabelAction.class) != null) {
                 if ((enclosing instanceof StepStartNode && ((StepStartNode) enclosing).getDescriptor() instanceof StageStep.DescriptorImpl) ||
                         (enclosing.getAction(ThreadNameAction.class) != null)) {
-                    enclosingBlocks.add(enclosing.getId());
+                    enclosingBlocks.add(enclosing);
                 }
             }
         }
 
         return enclosingBlocks;
     }
 
+    @Nonnull
+    public static List<String> getEnclosingBlockIds(@Nonnull List<FlowNode> nodes) {
+        List<String> ids = new ArrayList<>();
+        for (FlowNode n : nodes) {
+            ids.add(n.getId());
+        }
+        return ids;
+    }
+
+    @Nonnull
+    public static List<String> getEnclosingBlockNames(@Nonnull List<FlowNode> nodes) {
+        List<String> names = new ArrayList<>();
+        for (FlowNode n : nodes) {
+            ThreadNameAction threadNameAction = n.getAction(ThreadNameAction.class);
+            if (n.getAction(LabelAction.class) != null && threadNameAction != null) {
```

Comment:
Why do you need to check for `LabelAction` here?

Reply:
Good question. Can't remember if there was a reason...